src/6model/reprs/MVMSpeshCandidate.c

```diff
@@ -0,0 +1,366 @@
+    candidate->body.lexical_types = sg->lexical_types;
+
+    MVM_free(sc);
+
+    /* Try to JIT compile the optimised graph. The JIT graph hangs from
+     * the spesh graph and can safely be deleted with it. */
+    if (tc->instance->jit_enabled) {
+        MVMJitGraph *jg;
+        if (MVM_spesh_debug_enabled(tc))
+            jit_time = uv_hrtime();
+
+        jg = MVM_jit_try_make_graph(tc, sg);
+        if (jg != NULL) {
+            candidate->body.jitcode = MVM_jit_compile_graph(tc, jg);
+            MVM_jit_graph_destroy(tc, jg);
+        }
+    }
+
+    if (MVM_spesh_debug_enabled(tc)) {
+        char *after = MVM_spesh_dump(tc, sg);
+        end_time = uv_hrtime();
+        MVM_spesh_debug_printf(tc, "After:\n%s", after);
+        MVM_spesh_debug_printf(tc,
+            "Specialization took %" PRIu64 "us (total %" PRIu64"us)\n",
+            (spesh_time - start_time) / 1000,
+            (end_time - start_time) / 1000);
+
+        if (tc->instance->jit_enabled) {
+            MVM_spesh_debug_printf(tc,
+                "JIT was %ssuccessful and compilation took %" PRIu64 "us\n",
+                candidate->body.jitcode ? "" : "not ", (end_time - jit_time) / 1000);
+            if (candidate->body.jitcode) {
+                MVM_spesh_debug_printf(tc, "    Bytecode size: %" PRIu64 " byte\n",
+                                       candidate->body.jitcode->size);
+            }
+        }
+        MVM_spesh_debug_printf(tc, "\n========\n\n");
+        MVM_free(after);
+        fflush(tc->instance->spesh_log_fh);
+    }
+
+    /* Calculate work environment taking JIT spill area into account. */
+    calculate_work_env_sizes(tc, sg->sf, candidate);
+
+    /* Update spesh slots. */
+    candidate->body.num_spesh_slots = sg->num_spesh_slots;
+    candidate->body.spesh_slots     = sg->spesh_slots;
+
+    /* Claim ownership of allocated memory assigned to the candidate. */
+    sg->cand = candidate;
+    MVM_spesh_graph_destroy(tc, sg);
+
+    /* Create a new candidate list and copy any existing ones. Free memory
+     * using the FSA safepoint mechanism. */
+    spesh = p->sf->body.spesh;
+    new_candidate_list = MVM_fixed_size_alloc(tc, tc->instance->fsa,
+        (spesh->body.num_spesh_candidates + 1) * sizeof(MVMSpeshCandidate *));
+    if (spesh->body.num_spesh_candidates) {
+        size_t orig_size = spesh->body.num_spesh_candidates * sizeof(MVMSpeshCandidate *);
+        memcpy(new_candidate_list, spesh->body.spesh_candidates, orig_size);
+        MVM_fixed_size_free_at_safepoint(tc, tc->instance->fsa, orig_size,
+            spesh->body.spesh_candidates);
+    }
+    MVM_ASSIGN_REF(tc, &(spesh->common.header), new_candidate_list[spesh->body.num_spesh_candidates], candidate);
+    spesh->body.spesh_candidates = new_candidate_list;
+
+    /* May now be referencing nursery objects, so barrier just in case. */
+    //if (spesh->common.header.flags2 & MVM_CF_SECOND_GEN)
+        //MVM_gc_write_barrier_hit(tc, (MVMCollectable *)spesh);
```

Comment:
I think this can safely go away now; we know that we just allocated the `MVMSpeshCandidate` object in the nursery.